### ISSUE TITLE
[bug] Increase prose_tokens

### ISSUE THREAD

**Issue by dribagnac** [Avarok-Cybersecurity/atlas#328]
### Atlas version / docker tag

avarok/atlas-gb10:latest (2026-07-01)

### Model

nvidia/Qwen3.6-35B-A3B-NVFP4

### Hardware

1x GB10, single CPU

### Launch command

```bash
docker run -d --name atlas \
  --network host --gpus all --ipc=host \
  --restart unless-stopped  \
  -v ~/.cache/huggingface:/root/.cache/huggingface \
  avarok/atlas-gb10:latest \
  serve nvidia/Qwen3.6-35B-A3B-NVFP4 \
    --host 0.0.0.0 \
    --port 8000 \
    --max-seq-len 65536 \
    --kv-cache-dtype fp8 \
    --kv-high-precision-layers auto \
    --gpu-memory-utilization 0.85 \
    --scheduling-policy slai \
    --enable-prefix-caching \
    --speculative \
	--mtp-quantization nvfp4  \
    --num-drafts 2
```

### Expected behaviour

increase prose_tokens value,  I don't find any option for that ?

### Actual behaviour

Using Pi.dev (Pi agent) with Atlas on my DGX10.
Following error on Pi.dev : 
 Error: Model stopped because it reached the maximum output token limit. The response may be incomplete.

Atlas log:
2026-07-17T08:06:27.560274Z  WARN spark::scheduler::decode_logits_content: Inter-tool prose budget exhausted, ending response (rollback declined) prose_tokens=385 max=384 reason=StreamUnsafe

Is-there a way to increase prose_tokens ? I don't find any option for that ?

Tanks.
Regards
Daniel

### Additional context

Using Pi agent (pi.dev)
 Error: Model stopped because it reached the maximum output token limit. The response may be incomplete.

**Comment by dribagnac** [comment]
Issue is not linked to the model (I tested with some Qwen 3.5 or 3.6 variants).

Qwen-Coder-Next
docker run -d --name atlasCoder \
  --network host --gpus all --ipc=host \
  --restart unless-stopped  \
  -v ~/.cache/huggingface:/root/.cache/huggingface \
  avarok/atlas-gb10:latest \
  serve unsloth/Qwen3-Coder-Next-FP8 \
    --host 0.0.0.0 \
    --port 8000 \
    --max-seq-len 65536 \
    --kv-cache-dtype fp8 \
    --kv-high-precision-layers auto \
    --gpu-memory-utilization 0.90 \
    --scheduling-policy slai \
    --enable-prefix-caching \
	--tool-max-tokens 32768 \
	--max-thinking-budget 4096  \
	--tool-call-parser qwen3_coder  \
	--speculative \
    --num-drafts 1

**Comment by dribagnac** [comment]
Tested with the new docker image 2026-08-13T05:08
issue still present

2026-08-14T12:32:45.831476Z  WARN spark::scheduler::decode_logits_content: Inter-tool prose budget exhausted, ending response (rollback declined) prose_tokens=385 max=384 reason=StreamUnsafe

**Comment by tbraun96** [comment]
Thanks Daniel — confirmed, reproduced, and fixed in #510.

**Unblock yourself today, no upgrade needed.** `ATLAS_MAX_INTER_TOOL_PROSE` already works on the image you are running; it was simply undocumented, missing from `--help`, and unmentioned in the warning you found:

```
docker run -e ATLAS_MAX_INTER_TOOL_PROSE=8192 ... avarok/atlas-gb10:latest serve ...
```

Any u32; it overrides the per-model value.

**Root cause.** A July change already diagnosed this exact complaint and raised the default from 384 to 3072 — but it raised a constant that nothing reads. `WatchdogParams::from_behavior` always overwrites it with the atlas-kernels model-behavior value, whose default stayed 384, and only one model got an explicit override. Your models set none, so you got 384. That change's own regression test asserted the dead constant, which is why it stayed green while the bug survived.

**In the next image:** the default becomes 3072 from a single shared constant, there is a discoverable `--max-inter-tool-prose` flag (`0` disables the guard entirely), and the warning names the knobs.

**One thing we got wrong that is worth calling out:** on your code path the truncated response was labelled `finish_reason="stop"` — the API told your agent the answer was *complete* when we had in fact cut it off. That is why Pi.dev reported a generic max-token error. Watchdog cuts are now labelled `finish_reason="length"` consistently so a client can tell truncation from completion.

The guard itself stays for now — it prevents a real runaway where prose loops between tool calls and burns your whole `max_tokens` — but it is now yours to size. We are separately investigating whether it should exist at all: it exists partly because our tool grammar may not let the model emit a stop token on its own, which would be the actual thing to fix. If that holds, the budget goes away rather than getting bigger.

Sorry this cost you a bug report to discover. Your second comment — that it reproduces on the 2026-08-13 image — is what made it findable.
